router/core/src/main/scala/com/twitter/finagle/buoyant/DstBindingFactory.scala

```diff
@@ -126,12 +133,16 @@ object DstBindingFactory {
     boundMk: Mk[Dst.Bound, Req, Rsp] = Mk.identity[Dst.Bound, Req, Rsp],
     namer: NameInterpreter = DefaultInterpreter,
     statsReceiver: StatsReceiver = DefaultStatsReceiver,
-    capacity: Capacity = Capacity.default
-  ) extends DstBindingFactory[Req, Rsp] {
+    capacity: Capacity = Capacity.default,
+    bindingTimeout: BindingTimeout = BindingTimeout.default
+  )(implicit timer: Timer = DefaultTimer.twitter) extends DstBindingFactory[Req, Rsp] {
     private[this]type Cache[Key] = XXX_ServiceFactoryCache[Key, Req, Rsp]
 
     def apply(dst: Dst, conn: ClientConnection): Future[Service[Req, Rsp]] = dst match {
-      case path: Dst.Path => pathCache(path, conn)
+      case path: Dst.Path => pathCache(path, conn).raiseWithin(
+        bindingTimeout.timeout,
+        new TimeoutException(s"Binding ${path.path.show} exceeded ${bindingTimeout.timeout}")
+      )
```

Comment:
`Future` already special cases this for us:

```
  def raiseWithin(timer: Timer, timeout: Duration, exc: Throwable): Future[A] = {
    if (timeout == Duration.Top || isDefined)
      return this
```

The exception message contains the path so it can't be val'd on the class.  I expected that the exception would be pass-by-name, but it looks like it's not :(  
Another option would be to make it extend `NoStackTrace` to somewhat reduce allocation.  Or remove the path from the message.